### PR TITLE
Port Alpakka Xml

### PR DESCRIPTION
**Pull request by Arkatufus**
This should work now

### CODE REVIEW

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+    {
+        public string Text { get; }
+
+        protected TextEvent(string text)
+        {
+            Text = text;
+        }
+    }
+
+    public sealed class StartDocument : IParseEvent
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: since those are only "markers" they can be a singleton

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+    }
+
+    public sealed class EndDocument : IParseEvent
+    {
+    }
+
+    public sealed class StartElement : IParseEvent
+    {
+        public string LocalName { get; }
+        public Dictionary<string, string> Attributes { get; }
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: Should be immutable

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+                // thats why we offset the data so it fits into the right side of the buffer.
+                var offset = _bufferSize - bytes.Count;
+                _feeder.Position = offset;
+                _feeder.Write(bytes.ToArray(), 0, bytes.Count);
+                _feeder.Position = offset;
+            }
+
+            public override void OnUpstreamFinish()
+            {
+                if (_hasNext)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: what if _hasNext is true but _pendingEvent is not `NULL`  ?

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+
+        public override string ToString() => nameof(StreamingXmlParser);
+    }
+    #endregion
+
+    #region Coalesce
+    /// <summary>
+    /// Internal API. Use <see cref="Akka.Streams.Xml.Dsl.XmlParsing"/> instead.
+    /// </summary>
+    public sealed class Coalesce : GraphStage<FlowShape<IParseEvent, IParseEvent>>
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: You can use `SimpleLinearGraphStage<IParseEvent>` in cases where  `in` and `out` type are the same

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+            {
+                Pull(_stage.In);
+            }
+
+            public override void OnPush()
+            {
+                var parseEvent = Grab(_stage.In);
+
+                var t = parseEvent as TextEvent;
+                if (t != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe want to use c# 7 syntax ?  `if(parseEvent is TextEvent t)`

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+
+        public override string ToString() => nameof(Coalesce);
+    }
+    #endregion
+
+    #region Subslice
+    /// <summary>
+    /// Internal API. Use <see cref="Akka.Streams.Xml.Dsl.XmlParsing"/> instead.
+    /// </summary>
+    public sealed class Subslice : GraphStage<FlowShape<IParseEvent, IParseEvent>>
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: As above, use SimpleLinearGraphStage

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+                PartialMatch,
+                NoMatch
+            }
+
+            private readonly Subslice _stage;
+            private readonly Lazy<PassThrough> _passThrough;
+            private readonly Lazy<PartialMatch> _partialMatch;
+            private readonly Lazy<NoMatch> _noMatch;
+
+            private MatchState State
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: that looks a bit strange ;-)  maybe a method would be better

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+                            SetHandler(_stage.In, _partialMatch.Value);
+                            break;
+                        case MatchState.NoMatch:
+                            SetHandler(_stage.In, _noMatch.Value);
+                            break;
+                    }
+                }
+            }
+            private Stack<string> Expected { get; set; }
+            private Stack<string> MatchedSoFar { get; } = new Stack<string>();
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: why not fields ?

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 9]
```diff
@@ -0,0 +1,707 @@
...
+                    _stage = stage;
+                    _logic = logic;
+                }
+
+                public override void OnPush()
+                {
+                    var inEvent = _logic.Grab(_stage.In);
+
+                    var start = inEvent as StartElement;
+                    if (start != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+                    {
+                        _depth++;
+                        _logic.Push(_stage.Out, start);
+                        return;
+                    }
+
+                    var end = inEvent as EndElement;
+                    if (end != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 11]
```diff
@@ -0,0 +1,707 @@
...
+                    _stage = stage;
+                    _logic = logic;
+                }
+
+                public override void OnPush()
+                {
+                    var inEvent = _logic.Grab(_stage.In);
+
+                    var start = inEvent as StartElement;
+                    if (start != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 12]
```diff
@@ -0,0 +1,707 @@
...
+                        else
+                        {
+                            _logic.State = MatchState.NoMatch;
+                        }
+                        _logic.Pull(_stage.In);
+                        return;
+                    }
+
+                    var end = inEvent as EndElement;
+                    if (end != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 13]
```diff
@@ -0,0 +1,707 @@
...
+                    _stage = stage;
+                    _logic = logic;
+                }
+
+                public override void OnPush()
+                {
+                    var inEvent = _logic.Grab(_stage.In);
+
+                    var start = inEvent as StartElement;
+                    if (start != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 14]
```diff
@@ -0,0 +1,707 @@
...
+                    {
+                        _depth++;
+                        _logic.Pull(_stage.In);
+                        return;
+                    }
+
+                    var end = inEvent as EndElement;
+                    if (end != null)
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: maybe c# 7 syntax

File: Xml/src/Akka.Streams.Xml/Xml.cs [hunk 15]
```diff
@@ -0,0 +1,707 @@
...
+                }
+            }
+        }
+
+        #endregion
+
+        public Subslice(IImmutableList<string> path)
+        {
+            Shape = new FlowShape<IParseEvent, IParseEvent>(In, Out);
+            Path = path.ToList();
```
Comment (Silv3rcircl3) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: why creating a list ?

**Comment by Arkatufus** [comment]
Done

File: Xml/src/Akka.Streams.Xml/Xml.cs (hunk 1 above)
Comment (Arkatufus) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: done

File: Xml/src/Akka.Streams.Xml/Xml.cs (hunk 3 above)
Comment (Arkatufus) [inline Xml/src/Akka.Streams.Xml/Xml.cs]: that was a piece of overlooked leftover code, thanks for catching it (deleted)

**Pull request merged**